File: javascript/ch1/ex108.js
load('../functions.js');

sq = function (x) {
    return x * x;
};

cube = function (x) {
    return x * x * x;
};

improve = function (guess, x) {
    return (x / sq(guess) + 2 * guess) / 3.0;
};

is_good_enough = function (guess, x) {
    return Math.abs(cube(guess) - x) < (guess / 100000);
};

cbrt_iter = function (guess, x) {
    return is_good_enough(guess, x) ? guess : cbrt_iter(improve(guess, x), x);
};

cbrt = function (x) {
    return cbrt_iter(1.0, x);
};

var delta = 0.0001;
check_expect_float(cube(cbrt(27)), 27, delta);
check_expect_float(cube(cbrt(0.27)), 0.27, delta);

File: clojure/src/ch1/ex108.clj
(defn square [x]
  (* x x))

(defn cube [x]
  (* x x x))

(defn improve [guess x]
  (/ (+ (/ x (square guess)) (* 2 guess)) 3))

(defn good-enough? [guess x]
  (< (Math/abs (- (cube guess) x)) (/ guess 100000)))

(defn cbrt-iter [guess x]
  (if (good-enough? guess x) guess (cbrt-iter (improve guess x) x)))

(defn my-cbrt [x]
  (cbrt-iter 1.0 x))

(require
 'clojure.test)

(def delta 0.00001)
(clojure.test/is (< (Math/abs (- (my-cbrt 27) 3.0)) delta))
(clojure.test/is (< (Math/abs (- (my-cbrt 0.000027) 0.03)) delta))

